src/celeritas/geo/GeoMaterialParams.cc

```diff
@@ -28,6 +28,211 @@
+                              std::vector<Label>&& labels,
+                              std::vector<MaterialId> const& materials)
+{
+    CELER_EXPECT(materials.size() == labels.size());
+    CELER_EXPECT(std::all_of(
+        materials.begin(), materials.end(), [&mat_params](MaterialId m) {
+            return !m || m < mat_params.num_materials();
+        }));
+
+    std::unordered_map<Label, MaterialId> lab_to_id;
+    std::set<Label> duplicates;
+
+    // Remap materials to volume IDs using given volume names:
+    // build a map of volume name -> matid
+    for (auto idx : range(materials.size()))
+    {
+        if (!materials[idx])
+        {
+            // Skip volumes without matids
+            continue;
+        }
+
+        auto [prev, inserted]
+            = lab_to_id.insert({std::move(labels[idx]), materials[idx]});
+        if (!inserted)
+        {
+            duplicates.insert(prev->first);
+        }
+    }
+    CELER_VALIDATE(duplicates.empty(),
+                   << "geo/material coupling specified duplicate "
+                      "volume names: \""
+                   << join(duplicates.begin(), duplicates.end(), "\", \"")
+                   << '"');
+    return lab_to_id;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find a material ID from a volume ID.
+ */
+class MaterialFinder
+{
+  public:
+    MaterialFinder(GeoParams const& geo, MapLabelMatId const& materials)
+        : geo_{geo}, materials_{materials}
+    {
+    }
+
+    MaterialId operator()(VolumeId const& volume_id)
+    {
+        Label const& vol_label = geo_.id_to_label(volume_id);
+
+        // Hopefully user-provided and geo-provided volume labels match exactly
+        if (auto iter = materials_.find(vol_label); iter != materials_.end())
+        {
+            return iter->second;
+        }
+
+        if (mat_labels_.empty())
+        {
+            // Build user volume -> material mapping
+            this->build_mat_labels();
+        }
+
+        // Either:
+        // - user-provided vol labels have no extensions? (just names)
+        // - geometry volume labels are missing extensions (e.g. when
+        //   Geant4-derived volume names, ORANGE geometry names)
+        auto [start, stop] = mat_labels_.equal_range(vol_label.name);
+        if (start == stop)
+        {
+            // No materials match the volume label
+            return {};
+        }
+        if (std::distance(start, stop) == 1)
+        {
+            return start->second.second;
+        }
+
+        // Multiple labels match
+        CELER_LOG(warning)
+            << "Multiple materials match the volume '" << vol_label << "': "
+            << join_stream(
+                   start, stop, ", ", [](std::ostream& os, auto&& mliter) {
+                       os << mliter.second.second.unchecked_get();
+                   });
+        return start->second.second;
```

Comment:
Maybe I'm reading this wrong, but since from this point on we return `start->second.second` for the `MaterialId`, might be easier to read if we simply do:

```cpp
if (std::distance(start, stop) > 1)
{
    // Multiple labels match
    CELER_LOG(warning)
        << "Multiple materials match the volume '" << vol_label << "': "
        << join_stream(
                start, stop, ", ", [](std::ostream& os, auto&& mliter) {
                    os << mliter.second.second.unchecked_get();
                });
}
return start->second.second;
```

Reply:
Good point. Thanks! With the output @amandalund got though, it'll make even more sense to build a temporary 'set' and only print if more than one ID exists in the set.